ArduCopter/mode_ship_ops.cpp

```diff
@@ -0,0 +1,754 @@
+    pos_control->set_correction_speed_accel_xy(wp_nav->get_default_speed_xy(), wp_nav->get_wp_acceleration());
+
+    // initialize vertical speeds and acceleration
+    pos_control->set_max_speed_accel_z(wp_nav->get_default_speed_down(), wp_nav->get_default_speed_up(), wp_nav->get_accel_z());
+    pos_control->set_correction_speed_accel_z(wp_nav->get_default_speed_down(), wp_nav->get_default_speed_up(), wp_nav->get_accel_z());
+
+    // initialise velocity controller
+    pos_control->init_z_controller();
+    pos_control->init_xy_controller();
+
+    // initialise yaw
+    // auto_yaw.set_mode_to_default(false);
+
+    if (get_alt_hold_state(0.0f) == AltHold_Flying) {
+        // we are flying so we will initialise at the climb to RTL altitude.
+        set_state(SubMode::CLIMB_TO_RTL);
+        ship_takeoff = false;
+    } else {
+        // we are landed so we will initialise in the Final state.
+        set_state(SubMode::LAUNCH_RECOVERY);
+        ship_takeoff = true;
+    }
+
+    return true;
+}
+
+// returns a user-readable string for supplied submode
+const char *ModeShipOperation::state_name(SubMode mode) const
+{
+    switch (mode) {
+    case SubMode::CLIMB_TO_RTL:
+        return "ClimbToRTL";
+    case SubMode::RETURN_TO_PERCH:
+        return "ReturnToPerch";
+    case SubMode::PERCH:
+        return "Perch";
+    case SubMode::OVER_SPOT:
+        return "OverSpot";
+    case SubMode::LAUNCH_RECOVERY:
+        return "LaunchRecovery";
+    case SubMode::PAYLOAD_PLACE:
+        return "PayloadPlace";
+    }
+    return "?";
+}
+
+void ModeShipOperation::set_state(SubMode mode)
+{
+    gcs().send_text(MAV_SEVERITY_WARNING, "ShipOps: %s",
+                    state_name(mode));
+    _state = mode;
+}
+
+void ModeShipOperation::set_keep_out_zone_mode(KeepOutZoneMode keep_out_zone_mode)
+{
+    if (_keep_out_zone_mode != keep_out_zone_mode) {
+        _keep_out_zone_mode = keep_out_zone_mode;
+        switch (_keep_out_zone_mode) {
+        case KeepOutZoneMode::NO_ACTION:
+            break;
+        case KeepOutZoneMode::AVOID_KOZ:
+            gcs().send_text(MAV_SEVERITY_WARNING, "ShipOps: Avoiding KOZ");
+            break;
+        case KeepOutZoneMode::EXIT_KOZ:
+            gcs().send_text(MAV_SEVERITY_WARNING, "ShipOps: Exiting KOZ");
+            break;
+        }
+    }
+
+}
+
+void ModeShipOperation::set_approach_mode(ApproachMode approach_mode)
+{
+    _approach_mode = approach_mode;
+    switch (_approach_mode) {
+    case ApproachMode::LAUNCH_RECOVERY:
+        gcs().send_text(MAV_SEVERITY_WARNING, "ShipOps: Mode Launch/Recovery");
+        break;
+    case ApproachMode::PAYLOAD_PLACE:
+        gcs().send_text(MAV_SEVERITY_WARNING, "ShipOps: Mode Payload Place");
+        break;
+    }
+
+}
+
+void ModeShipOperation::run()
+{
+    float yaw_cd = attitude_control->get_att_target_euler_cd().z;
+    float yaw_rate_cds = 0.0f;
+
+    Vector2f perch_offset = Vector2f(ship_perch_radius * 100.0, 0.0f);
```

Comment:
```suggestion
    Vector2f perch_offset { ship_perch_radius * 100.0, 0.0f };
```
(this format adds some sanity checking that you're not trying to put bad values into the structure, rather than doing the "make it fit" thing that C/C++ usually do)

Reply:
Done